Changes in module 'nf-core/tiddit/sv'
--- modules/nf-core/tiddit/sv/main.nf
+++ modules/nf-core/tiddit/sv/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::tiddit=3.3.2"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/tiddit:3.3.2--py310hc2b7f4b_0' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/tiddit:3.3.2--py310hc2b7f4b_0' :
         'biocontainers/tiddit:3.3.2--py310hc2b7f4b_0' }"
 
     input:

************************************************************
