Changes in module 'nf-core/gatk4/markduplicatesspark'
--- modules/nf-core/gatk4/markduplicatesspark/main.nf
+++ modules/nf-core/gatk4/markduplicatesspark/main.nf
@@ -3,7 +3,9 @@
     label 'process_high'
 
     conda "bioconda::gatk4=4.4.0.0 conda-forge::openjdk=8.0.312"
-    container "nf-core/gatk:4.4.0.0"
+    container "${ task.ext.container_full_uri ?
+        'quay.io/nf-core/gatk:4.4.0.0' :
+        'nf-core/gatk:4.4.0.0' }"
 
     input:
     tuple val(meta), path(bam)

************************************************************
