Changes in module 'nf-core/gatk4/markduplicates'
--- modules/nf-core/gatk4/markduplicates/main.nf
+++ modules/nf-core/gatk4/markduplicates/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::gatk4=4.4.0.0 bioconda::samtools=1.17"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mulled-v2-d9e7bad0f7fbc8f4458d5c3ab7ffaaf0235b59fb:f857e2d6cc88d35580d01cf39e0959a68b83c1d9-0':
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/mulled-v2-d9e7bad0f7fbc8f4458d5c3ab7ffaaf0235b59fb:f857e2d6cc88d35580d01cf39e0959a68b83c1d9-0' :
         'biocontainers/mulled-v2-d9e7bad0f7fbc8f4458d5c3ab7ffaaf0235b59fb:f857e2d6cc88d35580d01cf39e0959a68b83c1d9-0' }"
 
     input:

************************************************************
