Changes in module 'nf-core/multiqc'
--- modules/nf-core/multiqc/main.nf
+++ modules/nf-core/multiqc/main.nf
@@ -1,16 +1,14 @@
 process MULTIQC {
-    label 'process_single'
+    label 'process_medium'
 
-    conda (params.enable_conda ? 'bioconda::multiqc=1.13' : null)
+    conda (params.enable_conda ? 'bioconda::multiqc=1.13a' : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/multiqc:1.13--pyhdfd78af_0' :
-        'quay.io/biocontainers/multiqc:1.13--pyhdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/multiqc:1.13a--pyhdfd78af_1' :
+        'quay.io/biocontainers/multiqc:1.13a--pyhdfd78af_1' }"
 
     input:
     path  multiqc_files, stageAs: "?/*"
-    path(multiqc_config)
-    path(extra_multiqc_config)
-    path(multiqc_logo)
+    tuple path(multiqc_config), path(multiqc_logo)
 
     output:
     path "*multiqc_report.html", emit: report
@@ -24,13 +22,11 @@
     script:
     def args = task.ext.args ?: ''
     def config = multiqc_config ? "--config $multiqc_config" : ''
-    def extra_config = extra_multiqc_config ? "--config $extra_multiqc_config" : ''
     """
     multiqc \\
         --force \\
+        $config \\
         $args \\
-        $config \\
-        $extra_config \\
         .
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
