Changes in module 'nf-core/unzip'
--- modules/nf-core/unzip/main.nf
+++ modules/nf-core/unzip/main.nf
@@ -5,6 +5,8 @@
     conda "conda-forge::p7zip=16.02"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/p7zip:16.02' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/p7zip:16.02' :
         'biocontainers/p7zip:16.02' }"
 
     input:

************************************************************
