Changes in module 'nf-core/cat/fastq'
--- modules/nf-core/cat/fastq/main.nf
+++ modules/nf-core/cat/fastq/main.nf
@@ -5,6 +5,8 @@
     conda "conda-forge::sed=4.7"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/ubuntu:20.04' :
+        task.ext.container_full_uri ?
+        'quay.io/nf-core/ubuntu:20.04' :
         'nf-core/ubuntu:20.04' }"
 
     input:

************************************************************
