Changes in module 'nf-core/bwamem2/mem'
--- modules/nf-core/bwamem2/mem/main.nf
+++ modules/nf-core/bwamem2/mem/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::bwa-mem2=2.2.1 bioconda::samtools=1.16.1"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mulled-v2-e5d375990341c5aef3c9aff74f96f66f65375ef6:2cdf6bf1e92acbeb9b2834b1c58754167173a410-0' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/mulled-v2-e5d375990341c5aef3c9aff74f96f66f65375ef6:2cdf6bf1e92acbeb9b2834b1c58754167173a410-0' :
         'biocontainers/mulled-v2-e5d375990341c5aef3c9aff74f96f66f65375ef6:2cdf6bf1e92acbeb9b2834b1c58754167173a410-0' }"
 
     input:

************************************************************
