Changes in module 'nf-core/bwamem2/index'
--- modules/nf-core/bwamem2/index/main.nf
+++ modules/nf-core/bwamem2/index/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::bwa-mem2=2.2.1"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/bwa-mem2:2.2.1--he513fc3_0' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/bwa-mem2:2.2.1--he513fc3_0' :
         'biocontainers/bwa-mem2:2.2.1--he513fc3_0' }"
 
     input:

************************************************************
