Changes in module 'nf-core/vcftools'
--- modules/nf-core/vcftools/main.nf
+++ modules/nf-core/vcftools/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::vcftools=0.1.16"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/vcftools:0.1.16--he513fc3_4' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/vcftools:0.1.16--he513fc3_4' :
         'biocontainers/vcftools:0.1.16--he513fc3_4' }"
 
     input:

************************************************************
