Changes in module 'nf-core/tabix/bgziptabix'
--- modules/nf-core/tabix/bgziptabix/main.nf
+++ modules/nf-core/tabix/bgziptabix/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::tabix=1.11"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/tabix:1.11--hdfd78af_0' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/tabix:1.11--hdfd78af_0' :
         'biocontainers/tabix:1.11--hdfd78af_0' }"
 
     input:

************************************************************
