Changes in module 'nf-core/dragmap/align'
--- modules/nf-core/dragmap/align/main.nf
+++ modules/nf-core/dragmap/align/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda (params.enable_conda ? "bioconda::dragmap=1.3.0 bioconda::samtools=1.16.1 conda-forge::pigz=2.3.4" : null)
+    conda (params.enable_conda ? "bioconda::dragmap=1.2.1 bioconda::samtools=1.15.1 conda-forge::pigz=2.3.4" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mulled-v2-580d344d9d4a496cd403932da8765f9e0187774d:bb7e33c2ede38421063b8379a5fb3453645bb047-0':
-        'quay.io/biocontainers/mulled-v2-580d344d9d4a496cd403932da8765f9e0187774d:bb7e33c2ede38421063b8379a5fb3453645bb047-0' }"
+        'https://depot.galaxyproject.org/singularity/mulled-v2-580d344d9d4a496cd403932da8765f9e0187774d:5ebebbc128cd624282eaa37d2c7fe01505a91a69-0':
+        'quay.io/biocontainers/mulled-v2-580d344d9d4a496cd403932da8765f9e0187774d:5ebebbc128cd624282eaa37d2c7fe01505a91a69-0' }"
 
     input:
     tuple val(meta) , path(reads)

************************************************************
