Changes in module 'nf-core/deepvariant'
--- modules/nf-core/deepvariant/main.nf
+++ modules/nf-core/deepvariant/main.nf
@@ -2,7 +2,9 @@
     tag "$meta.id"
     label 'process_high'
 
-    container "nf-core/deepvariant:1.5.0"
+    container "${ task.ext.container_full_uri ?
+        'quay.io/nf-core/deepvariant:1.5.0' :
+        'nf-core/deepvariant:1.5.0' }"
 
     // Exit if running this module with -profile conda / -profile mamba
     if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {

************************************************************
