Changes in module 'nf-core/ascat'
--- modules/nf-core/ascat/main.nf
+++ modules/nf-core/ascat/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::ascat=3.1.1 bioconda::cancerit-allelecount=4.3.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mulled-v2-c278c7398beb73294d78639a864352abef2931ce:ba3e6d2157eac2d38d22e62ec87675e12adb1010-0':
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/mulled-v2-c278c7398beb73294d78639a864352abef2931ce:ba3e6d2157eac2d38d22e62ec87675e12adb1010-0' :
         'biocontainers/mulled-v2-c278c7398beb73294d78639a864352abef2931ce:ba3e6d2157eac2d38d22e62ec87675e12adb1010-0' }"
 
     input:

************************************************************
