Changes in module 'nf-core/mosdepth'
--- modules/nf-core/mosdepth/main.nf
+++ modules/nf-core/mosdepth/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::mosdepth=0.3.3"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mosdepth:0.3.3--hdfd78af_1' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/mosdepth:0.3.3--hdfd78af_1' :
         'biocontainers/mosdepth:0.3.3--hdfd78af_1'}"
 
     input:

************************************************************
