Changes in module 'nf-core/fgbio/callmolecularconsensusreads'
--- modules/nf-core/fgbio/callmolecularconsensusreads/main.nf
+++ modules/nf-core/fgbio/callmolecularconsensusreads/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::fgbio=2.0.2"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/fgbio:2.0.2--hdfd78af_0' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/fgbio:2.0.2--hdfd78af_0' :
         'biocontainers/fgbio:2.0.2--hdfd78af_0' }"
 
     input:

************************************************************
