Changes in module 'nf-core/controlfreec/assesssignificance'
--- modules/nf-core/controlfreec/assesssignificance/main.nf
+++ modules/nf-core/controlfreec/assesssignificance/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_low'
 
-    conda "bioconda::control-freec=11.6b"
+    conda "bioconda::control-freec=11.6"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/control-freec:11.6b--hdbdd923_0' :
-        'biocontainers/control-freec:11.6b--hdbdd923_0' }"
+        'https://depot.galaxyproject.org/singularity/control-freec:11.6--h1b792b2_1' :
+        'biocontainers/control-freec:11.6--h1b792b2_1' }"
 
     input:
     tuple val(meta), path(cnvs), path(ratio)

************************************************************
