Changes in module 'nf-core/bcftools/stats'
--- modules/nf-core/bcftools/stats/main.nf
+++ modules/nf-core/bcftools/stats/main.nf
@@ -8,10 +8,8 @@
         'quay.io/biocontainers/bcftools:1.15.1--h0ea216a_0' }"
 
     input:
-    tuple val(meta), path(vcf), path(tbi)
-    path regions
-    path targets
-    path samples
+    tuple val(meta), path(vcf)
+    path(target_bed)
 
     output:
     tuple val(meta), path("*stats.txt"), emit: stats
@@ -23,15 +21,11 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def regions_file = regions ? "--regions-file ${regions}" : ""
-    def targets_file = targets ? "--targets-file ${targets}" : ""
-    def samples_file =  samples ? "--samples-file ${samples}" : ""
+    def target = target_bed ? "--regions-file ${target_bed}" : ""
     """
     bcftools stats \\
         $args \\
-        $regions_file \\
-        $targets_file \\
-        $samples_file \\
+        $target \\
         $vcf > ${prefix}.bcftools_stats.txt
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
