Changes in module 'nf-core/cnvkit/antitarget'
--- modules/nf-core/cnvkit/antitarget/main.nf
+++ modules/nf-core/cnvkit/antitarget/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::cnvkit=0.9.9 bioconda::samtools=1.16.1"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/cnvkit:0.9.9--pyhdfd78af_0':
+        task.ext.container_full_uri ?
+        'quay.io/cnvkit:0.9.9--pyhdfd78af_0' :
         'biocontainers/cnvkit:0.9.9--pyhdfd78af_0' }"
 
     input:

************************************************************
