Changes in module 'nf-core/bcftools/annotate'
--- modules/nf-core/bcftools/annotate/main.nf
+++ modules/nf-core/bcftools/annotate/main.nf
@@ -8,7 +8,10 @@
         'biocontainers/bcftools:1.20--h8b25389_0' }"
 
     input:
-    tuple val(meta), path(input), path(index), path(annotations), path(annotations_index), path(header_lines)
+    tuple val(meta), path(input)
+    path annotations
+    path annotations_index
+    path header_lines
 
     output:
     tuple val(meta), path("*.{vcf,vcf.gz,bcf,bcf.gz}"), emit: vcf
@@ -29,6 +32,10 @@
                     "vcf"
     if ("$input" == "${prefix}.${extension}") error "Input and output names are the same, set prefix in module configuration to disambiguate!"
     """
+    bcftools \\
+        index \\
+        $input
+
     bcftools \\
         annotate \\
         $args \\

************************************************************
