Changes in module 'nf-core/bwa/index'
--- modules/nf-core/bwa/index/main.nf
+++ modules/nf-core/bwa/index/main.nf
@@ -5,6 +5,8 @@
     conda "bioconda::bwa=0.7.17"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/bwa:0.7.17--hed695b0_7' :
+        task.ext.container_full_uri ?
+        'quay.io/biocontainers/bwa:0.7.17--hed695b0_7' :
         'biocontainers/bwa:0.7.17--hed695b0_7' }"
 
     input:

************************************************************
