Changes in module 'nf-core/sentieon/dedup'
--- modules/nf-core/sentieon/dedup/main.nf
+++ modules/nf-core/sentieon/dedup/main.nf
@@ -24,6 +24,7 @@
     tuple val(meta), path("*.bai"),     emit: bai
     tuple val(meta), path("*.score"),   emit: score
     tuple val(meta), path("*.metrics"), emit: metrics
+    tuple val(meta), path("*.metrics.multiqc.tsv"), emit: metrics_multiqc_tsv
     path "versions.yml",                emit: versions
 
     when:
@@ -60,6 +61,9 @@
 
     sentieon driver $args $input_list -r ${fasta} --algo LocusCollector $args2 --fun score_info ${prefix}.score
     sentieon driver $args3 -t $task.cpus $input_list -r ${fasta} --algo Dedup $args4 --score_info ${prefix}.score --metrics ${metrics} ${prefix}${suffix}
+    # This following tsv-file is produced in order to get a proper tsv-file with Dedup-metrics for importing in MultiQC as "custom content".
+    # It should be removed once MultiQC has a module for displaying Dedup-metrics.
+    head -3 ${metrics} > ${metrics}.multiqc.tsv
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
